### PR TITLE
Trusted external command

### PR DESCRIPTION
**Pull request by reidmv**
This lets PE read data from Satellite, saving it to `trusted.external.satellite`.

This capability deepens PE's integration with Satellite 6, and is a first step in potential migration paths for current Satellite 6 + Puppet customers who are today using Satellite 6's included Puppet server, who are migrating to Satellite 7 (which doesn't include Puppet), and want to keep using the Satellite interface and workflow while moving Puppet functionality over to PE.

### CODE REVIEW

**Comment by chelnak** [comment]
Hey @reidmv, thank you for this.

It looks like this module might need a bit of TLC so I've raised it with the team and we will try to clean things up a bit so that it makes the integration of this PR cleaner!

**Review by davidsandilands — APPROVED** [review]
Reviewed and approved from a Solutions architecture point of view awaiting review from module owner

**Comment by chelnak** [comment]
@reidmv @davidsandilands I've given this a bump internally. Hopefully we can get some eyes on it shortly.

File: manifests/init.pp
Comment (ekohl) [inline manifests/init.pp:99]: Wouldn't it make sense to have root own this? When would Puppet ever write to this directory at runtime? Feels like a security risk.

File: manifests/init.pp
Comment (ekohl) [inline manifests/init.pp:115]: Same here

File: manifests/init.pp
Comment (ekohl) [inline manifests/init.pp:117]: I've been getting using to using `content => file("${module_name}/satellite")` so it's inlined in the catalog.

File: files/satellite
Comment (ekohl) [inline files/satellite:18]:
```suggestion
```

File: README.md
Comment (ekohl) [inline README.md:80]: The produced RPMs are deprecated and will be dropped in a future release.

File: README.md
Comment (ekohl) [inline README.md:103]: I'm not sure this combination makes sense. If you restrict it to smart proxies, the fact it's a smart proxy is sufficient. So this should be sufficient if the CN on the client certificate is also registered as a Smart Proxy:
```suggestion
  6. In the Satellite UI, go to *Administer -> Settings -> Auth* and set the `restrict_registered_smart_proxies` parameter to `true`.
```

File: manifests/init.pp
Comment (ekohl) [inline manifests/init.pp:95]: Would it make sure to ensure the integration files are absent in the `else` case?

**Review by ekohl — COMMENTED** [review]
I realize we have a chat scheduled for Friday so feel free to consider my feedback preparation.

Using this new approach you wouldn't have the global variables only under `$trusted['external']['satellite']['parameters']`, right?

File: manifests/init.pp
Comment (reidmv) [reply manifests/init.pp:117]: Curious, what would be the benefit of inlining this kind of file?

My instinct is to avoid inlining, since it's a static non-templated file with a fair bit of content to it. Static catalogs already inlines the checksum information, so the agent will only request a copy from the server in a separate API call if it _knows_ that it needs to replace it; most agent runs will read the inlined checksum, see that it matches, and proceed. Inlining the actual content would make the catalog larger, but not provide runtime performance improvements for the 99% case.

Is there another reason why inlining might be good here, or in general?

File: README.md
Comment (reidmv) [reply README.md:80]: Until line 121, README changes in this commit are just for markdown formatting. I do not mean to suggest any content changes in this PR outside of the additions starting on line 121.

File: README.md
Comment (reidmv) [reply README.md:103]: See comment above; I do not mean to suggest content changes here, it's just a markdown formatting fix.

File: manifests/init.pp
Comment (reidmv) [reply manifests/init.pp]: It's a little tricky. I've updated the parameter name to `$manage_trusted_external_command` (which matches with the other Boolean parameter in the class) which I think should reduce the intuition that there is any de-configuration being offered here. Does that still feel like it needs a de-configuration option?

Part of the trickiness is in management of the surrounding config; namely, the creation of the directory (which should be a Puppet standard thing, not owned by satellite\_pe\_tools), and the management of the puppet.conf setting. I kind of want to avoid the mess of stepping into that in situations where satellite is not the only trusted external command.

File: manifests/init.pp
Comment (ekohl) [reply manifests/init.pp:117]: Some of this may be outdated but back in the day it was faster for small files. Source means a HTTP GET which does generate some additional load. There is (by now) an ancient blog which benchmarked it and found that up to some size (IIRC 4k) it was faster to inline it. You may be right that by now puppetserver learned optimizations that remove the benefit.

To me now it mostly means that catalog compilation checks if the path is actually present. That means that rspec-puppet can ensure the source is available rather than checking it at runtime.

Another small thing may be that a cached catalog can be applied which could happen if the server is offline. I'm not sure how practical that is.

There is some downside: if you support `ensure => absent` the `content => file()` is still executed which is technically redundant.

Overall I like the pattern a lot more and rarely use `source =>` at all.

File: README.md
Comment (ekohl) [reply README.md:80]: Consider it a talking point for our chat.

**Comment by reidmv** [comment]
> Using this new approach you wouldn't have the global variables only under `$trusted['external']['satellite']['parameters']`, right?

Correct. We're exposing everything that the Satellite ENC endpoint returns, but not in the global namespace. It's only available under `getvar('trusted.external.satellite')`. E.g. `getvar('trusted.external.satellite.parameters.organization')`.

```
{
      "classes" : {
        "stdlib::stages" : null
      },
      "environment" : "production",
      "parameters" : {
        "content_view" : "Default_Organization_View",
        "content_view_info" : {
          "components" : { },
          "label" : "Default_Organization_View",
          "latest-version" : "1.0",
          "published" : "2022-03-30 00:24:21 UTC",
          "version" : "1.0"
        },
        "domainname" : "c.puppet-solutions-architects.internal",
        "enable-epel" : "false",
        "enable-puppet5" : "true",
        "foreman_config_groups" : [ ],
        "foreman_env" : "production",
        "foreman_host_collections" : [ ],
        "foreman_interfaces" : [ {
          "attached_to" : null,
          "attrs" : { },
          "identifier" : "eth0",
          "ip" : "10.138.0.10",
          "ip6" : null,
          "link" : true,
          "mac" : "42:01:0a:8a:00:0a",
          "managed" : true,
          "name" : "rhel-node-1.c.puppet-solutions-architects.internal",
          "primary" : true,
          "provision" : true,
          "subnet" : null,
          "subnet6" : null,
          "tag" : null,
          "type" : "Interface",
          "virtual" : false
        } ],
        "foreman_subnets" : [ ],
        "foreman_users" : {
          "admin" : {
            "description" : null,
            "firstname" : "Admin",
            "fullname" : "Admin User",
            "lastname" : "User",
            "mail" : "root@c.puppet-solutions-architects.internal",
            "name" : "admin",
            "ssh_authorized_keys" : [ ]
          }
        },
        "host_registration_insights" : true,
        "host_registration_remote_execution" : true,
        "hostgroup" : "RHEL Nodes",
        "kickstart_repository" : null,
        "kt_cv" : "Default_Organization_View",
        "kt_env" : "Library",
        "lifecycle_environment" : "Library",
        "location" : "Default Location",
        "location_title" : "Default Location",
        "organization" : "Default Organization",
        "organization_title" : "Default Organization",
        "owner_email" : "root@c.puppet-solutions-architects.internal",
        "owner_name" : "Admin User",
        "puppet_ca" : "satellite-1.c.puppet-solutions-architects.internal",
        "puppetmaster" : "satellite-1.c.puppet-solutions-architects.internal",
        "remote_execution_connect_by_ip" : false,
        "remote_execution_effective_user_method" : "sudo",
        "remote_execution_ssh_keys" : [ "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDQXlRjK7vIweBbTJa5AtGguo5XOGtson88WTT9Nom4pJKZM7bJ+v1Z3nxeIhojdDZBTvF88GX6xqKsfyAWsfsciv/3pSjm7vUpg2rUndua9aIDIp40uqEyRNK9UXf8BF25IHBj3pTNn6KQ137DhZy0oXj4ezpUe3UB7QfXoHndPzP5ZicwN/O2VrgDaOznGtz3bZeDD/vNW6CZKA0HXmzl5ICLwz/rWPVBXO8EcUFhWo1dovTna6TOOdaRffBirNETMPF/x/YrSCBAxn7VvRXez80bod41CL88SD5RnC/qQb0ajlkU8A3qF83BJbevL8tVAdHQEGzxrUGbEKAJANm5 foreman-proxy@satellite-1" ],
        "remote_execution_ssh_user" : "root",
        "root_pw" : null,
        "ssh_authorized_keys" : [ ]
      }
}
```

File: manifests/init.pp
Comment (ekohl) [reply manifests/init.pp]: > Part of the trickiness is in management of the surrounding config; namely, the creation of the directory (which should be a Puppet standard thing, not owned by satellite_pe_tools), and the management of the puppet.conf setting. I kind of want to avoid the mess of stepping into that in situations where satellite is not the only trusted external command.

I get that, which is why I only suggested to ensure the specific command and config file would be absent. Perhaps the directory resource should use `ensure_resource()` to deal with duplication definitions.

Comment (reidmv) [reply manifests/init.pp]: Hmm. That's an interesting idea. I'm not a fan of `ensure_resource()`, generally speaking. I find that it introduces a Prisoner's Dilemma-like situation to your Puppet code, and is therefore never an ideal approach to a cross-module problem.

I think a longer-term fix which avoids a "kicking the can down the road" feel to it would be to define the dependent configuration in an appropriate upstream module. For PE, that would be in puppetlabs/puppet_enterprise. There is no reason the setting and directory should ever be anything different from the values given here. At that point, we would simply remove the resources from this module.

The only resource that belongs here is `File[/etc/puppetlabs/puppet/trusted-external-commands/satellite]`.

File: manifests/init.pp
Comment (reidmv) [reply manifests/init.pp:117]: Yeah, there is no longer any additional load from `source` as long as you have static catalogs enabled (by default in PE).

The rspec-puppet utility is interesting. I haven't thought about that one before.

My preference is still to use `source` when appropriate in order to keep the catalog smaller. But preferences like that are the module maintainer's call, and if as the maintainer you get a utility value out of using `content` instead, that's fine by me. 👍

**Comment by reidmv** [comment]
@chelnak after a discussion with @ekohl, it sounds like the best place to land these features might be in a module focused on integrating Puppet with Foreman, such as [theforeman/puppetserver_foreman](https://forge.puppet.com/modules/theforeman/puppetserver_foreman), and making that module compatible with both OSP and PE. Such a path would either be a successor/replacement for satellite_pe_tools, or if not that, could become a simple dependency of satellite_pe_tools.

We should talk about how best to align on continued support for Puppet users who also use Satellite, and how/if satellite_pe_tools fits into that picture.

**Comment by chelnak** [comment]
Sounds good to me 😄. Would be worth looping in @Aaron-Rob and Micah in to that conversation too.

**Comment by bastelfreak** [comment]
@chelnak https://tickets.puppetlabs.com/browse/SOLARCH-883 isn't a public ticket, could you make that public/move to a different board?

**Comment by chelnak** [comment]
@bastelfreak Maybe @davidsandilands can answer that one for you.

Regarding this ticket, my understanding is that this suggestion is now redundant. 

A change is due to be raised in the respective puppet_foreman module soon.

**Comment by davidsandilands** [comment]
@bastelfreak i don't think we have a good way of doing that unless @nigelkersten knows differently but I will mark into the ticket to put a link here for when @reidmv raises a PR against [puppet-puppetserver_foreman](https://github.com/theforeman/puppet-puppetserver_foreman)

**Pull request closed without merging**
